### PR TITLE
Print Functionality in Online POS using allow print before pay check

### PR DESCRIPTION
**Pull request by deepak-mnt**
1. "_Allow print before pay_" functionality in Online POS
2. Test Cases to check the print before pay functionality in Online POS
3. Handles the Sales Invoice update/creation in both cases(print before and after submit)

jmi_pr_1

### CODE REVIEW

File: erpnext/selling/page/point_of_sale/point_of_sale.js [hunk 1]
```diff
@@ -50,11 +50,11 @@ erpnext.pos.PointOfSale = class PointOfSale {
 				this.set_online_status();
 			},
 			() => this.setup_company(),
-
 			() => this.make_new_invoice(),
 			() => {
 				frappe.dom.unfreeze();
 			},
+			() => this.fetch_pos_profile_data(),
```
Comment (rohitwaghchaure) [inline erpnext/selling/page/point_of_sale/point_of_sale.js]: No need to write extra method, use existing method.
Set value in the method set_pos_fields, you can add the field allow_print_before_pay in the sales invoice and use same in the POS.
https://github.com/frappe/erpnext/blob/develop/erpnext/accounts/doctype/sales_invoice/sales_invoice.py#L293

File: erpnext/selling/page/point_of_sale/point_of_sale.js [hunk 2]
```diff
@@ -50,8 +50,8 @@ erpnext.pos.PointOfSale = class PointOfSale {
 				this.set_online_status();
 			},
 			() => this.setup_company(),
-
 			() => this.make_new_invoice(),
+			() => this.set_form_action(),
```
Comment (tundebabzy) [inline erpnext/selling/page/point_of_sale/point_of_sale.js]: Let's hide the print button until there is actually an item in the cart.
peek 2018-01-09 20-30

**Review by tundebabzy — CHANGES_REQUESTED** [review]

File: erpnext/selling/page/point_of_sale/point_of_sale.js (hunk 2 above)
Comment (saurabh-bhosale) [reply erpnext/selling/page/point_of_sale/point_of_sale.js]: @tundebabzy Thanks for the input. Have made the changes according to the suggestions in new commits.

File: erpnext/selling/page/point_of_sale/point_of_sale.py [hunk 3]
```diff
@@ -88,11 +88,15 @@ def get_conditions(item_code, serial_no, batch_no, barcode):
...
 
 @frappe.whitelist()
-def submit_invoice(doc):
+def submit_invoice(doc,islocal_dummy):
 	if isinstance(doc, basestring):
 		args = json.loads(doc)
 
-	doc = frappe.new_doc('Sales Invoice')
+	if islocal_dummy:
+		doc = frappe.get_doc('Sales Invoice',args["name"])
```
Comment (tundebabzy) [inline erpnext/selling/page/point_of_sale/point_of_sale.py]: Looks like this is causing problems with saving
peek 2018-01-12 12-54

**Review by tundebabzy — CHANGES_REQUESTED** [review]

File: erpnext/selling/page/point_of_sale/point_of_sale.py (hunk 3 above)
Comment (saurabh-bhosale) [reply erpnext/selling/page/point_of_sale/point_of_sale.py]: @tundebabzy  Fixed the bug and refactored the code a bit.

jmi_pr_1_rev

File: erpnext/selling/page/point_of_sale/point_of_sale.js [hunk 4]
```diff
@@ -499,19 +504,26 @@ erpnext.pos.PointOfSale = class PointOfSale {
...
-			this.frm.print_preview.printit(true);
-		});
-
-		this.page.set_primary_action(__("New"), () => {
-			this.make_new_invoice();
-		});
+		if(this.frm.doc.docstatus == 1 || (this.frm.doc.allow_print_before_pay == 1&&this.frm.doc.items.length>0)){
+			this.page.set_secondary_action(__("Print"), async() => {
+				if(this.frm.doc.docstatus != 1 ){
+					await this.frm.save();
```
Comment (tundebabzy) [inline erpnext/selling/page/point_of_sale/point_of_sale.js]: I don't know if async/await is ready for mainstream yet

**Review by tundebabzy — COMMENTED** [review]
Works as expected. @netchampfaris is IE a supported browser?

File: erpnext/selling/page/point_of_sale/point_of_sale.js (hunk 4 above)
Comment (tundebabzy) [reply erpnext/selling/page/point_of_sale/point_of_sale.js]: Ok. I just checked caniuse. I guess this is ok as long as we are not concerned about IE

File: erpnext/accounts/doctype/sales_invoice/sales_invoice.json [hunk 5]
```diff
@@ -297,6 +297,36 @@
...
+   "print_hide": 1, 
+   "print_hide_if_no_value": 0, 
+   "read_only": 1, 
+   "remember_last_selected_value": 0, 
+   "report_hide": 0, 
+   "reqd": 0, 
+   "search_index": 0, 
+   "set_only_once": 0, 
+   "unique": 0
+  },
```
Comment (rohitwaghchaure) [inline erpnext/accounts/doctype/sales_invoice/sales_invoice.json]: Why modified date has not changed even after you have added new field, can you rebase?

**Comment by rohitwaghchaure** [comment]
@deepak-mnt 
Closing the PR, kindly make changes and reopen it

**Pull request closed without merging**